More examples, extended HTTP client to send custom headers

sompylasar opened the pull request:
- Added support for request chaining.
  - Added AST transform to wrap function call arguments into promises.
  - Added more comments.
- Added a cerealizable-like `DemoObject` for the JSON-related examples.
- Extended the HTTP client API to send custom headers.
  - @dkorolev Bricks API should reflect this extension. I'd also rename `HTTPHeaders().Set()` to `HTTPHeaders().Add()` (according to its actual behavior) - not renamed in this PR yet.
- Added more examples to the README.
- Added more tests.

sompylasar committed:
Added DemoObject to the REPL context.

sompylasar committed:
README: Added Node.js requirement, JSON examples.

sompylasar committed:
Added support for request chaining.

* Added AST transform to wrap function call arguments into promises.
* Added more comments.

sompylasar committed:
README: Minor punctuation.

sompylasar committed:
README: Minor fix.

sompylasar committed:
Extended the HTTP client API to send custom headers

Added more tests for the new features and for the older ones.

sompylasar committed:
Added example for POST custom headers

dkorolev commented on README.md:
```diff
@@ -1,8 +1,10 @@
 # KnowSheet-REPL
-REPL interactive shell that demonstrates Bricks C++ syntax.
+A Read-Eval-Print-Loop (REPL) to demonstrate the KnowSheet Bricks C++ syntax.
```
How about "A Read-Eval-Print-Loop (REPL) _shell_ ..." ? :-)

sompylasar replied:
OK, will add.

dkorolev commented on README.md:
````diff
@@ -17,6 +19,18 @@ npm start
+echo '(expression-here)' | npm start
+```
+
+... and use the result in another command:
````
`s/and/or/` ?

sompylasar replied:
It was written `or` first :) then I thought that this line is also "evaluate one expression from the command-line".

dkorolev commented on README.md:
````diff
@@ -17,6 +19,18 @@ npm start
 
+... or evaluate one expression from the command-line:
+```bash
+echo '(expression-here)' | npm start
````
Why don't we change the `(expression-here)` into something useful?

How about some `JSONParse(HTTP(GET("knowsheet.io/uptime")).body).uptime.uptime_total_d`? I know it's long and not yet working, but the latter can/will be fixed, and ... you got the idea :)

sompylasar replied:
I thought that introducing an expression here before actually listing available expressions would be confusing.

dkorolev commented on README.md:
````diff
@@ -27,6 +41,17 @@ Perform an HTTP POST request and print the response:
+```
+KnowSheet> JSONParse(HTTP(POST("http://httpbin.org/post", DemoObject())).body).json
+```
+<sup>(in C++, you'll have a cerealizable object in place of the `DemoObject()`)</sup>
````
I'd stress that the C++ syntax is the same. Ex. <sup>The syntax mimics C++ Bricks exactly, as long as `DemoObject` is defined as a serializable type.</sup>.

sompylasar replied:
OK.

dkorolev commented on README.md:
````diff
@@ -37,9 +62,13 @@ Provide a custom User-Agent:
+Provide extra HTTP headers:
+```
+KnowSheet> HTTP(GET("http://httpbin.org/get?query=1", HTTPHeaders().Set("Custom", "Header"))).body
+KnowSheet> HTTP(POST("http://httpbin.org/post", "BODY", "text/plain", HTTPHeaders().Set("Custom", "Header"))).body
````
We can use multi-line commands like in node's REPL shell, right? Why not break this line then?

sompylasar replied:
I tried to make this one multiline, but when copy-pasted, the REPL executed it line-by-line giving errors on incomplete lines and partial success on complete lines.

dkorolev commented on lib/bricks-evaluate.js:
```diff
@@ -49,6 +49,24 @@ function createContext() {
 	extend(context, require('./bricks-net-api'));
 	
 	
+	// Add a `DemoObject` to demonstrate the JSON-POST.
```
Nit: Please don't use "string" as the field name, since it's going to get C++ a bit crazy. Thanks!

sompylasar replied:
Oh, right.

dkorolev commented on lib/bricks-evaluate.js:
```diff
@@ -49,6 +49,24 @@ function createContext() {
+			}
+		};
+	};
+	context.DemoObject.inspect = context.DemoObject.toString = function () {
```
A comment here would be nice. Is this just for the test?

sompylasar replied:
The comment comes above to the whole class. These lines work if you evaluate just `DemoObject`, e.g.:

```
KnowSheet> DemoObject
// A cerealizable class for JSON-POST demonstration.
```

dkorolev replied:
Very cute!

dkorolev commented on test/bricks-net-api.test.js:
```diff
@@ -631,4 +769,30 @@ describe('bricks-net-api', function () {
+			when(
+				api.JSON(object)
+			).then(function (result) {
+				assert.strictEqual(JSON.stringify(object), result);
```
Nit: It's a bit better to use plain strings in unit tests, ex. `assert.strictEqual('{"foo":"bar"}', ...`.

Not really important for this particular one ... although, since there is a nonzero chance someone hacking this code later might think it's a good idea to override `JSON.stringify()`, I'd say it's still worth applying to even this very test, since the thread is not entirely hypothetical.

(Same goes for things like HTTP error codes, where in test it's preferred to use the numeric constants, etc.)

sompylasar replied:
OK. I thought of this, too.

dkorolev commented:
Thanks, LGTM++!

A few very minor comments; the code one are exclusively nits, and the README ones are safe to be taken care of in the consecutive pull requests.

sompylasar committed:
CR: Fix for the intro sentence.

sompylasar committed:
CR: Fix for the CLI examples.

sompylasar committed:
CR: Fixed the CLI example phrase.

sompylasar committed:
CR: Fix for the DemoObject example.

sompylasar committed:
CR: Fixed the DemoObject property names for C++ compatibility.

sompylasar committed:
CR: Fixed the docu for the custom inspect functions.

Changed the `DemoObject` to create instances of its class.

sompylasar committed:
CR: Replaced JSON.stringify with constant strings in the tests.

sompylasar committed:
cpp-arguments: Fixed empty arguments with callbacks case.

sompylasar committed:
URL: Fixed empty arguments case, added test for it.

sompylasar committed:
REPL: Fixed and improved auto-completion.

sompylasar commented:
@dkorolev Fixed most of the nits. Added custom code completion because the default one gave completions that are invalid for Bricks. Fixed several discovered bugs (see the commits).  Please review once again.

dkorolev commented:
Yay for auto-completion!

One really minor comment. And one feature request: Is it possible to configure the prompt somehow?

Thanks,
Dima

sompylasar committed:
Added default prompt override via command line argument.

Changed `npm start` to `node run` to have a single command to start the
program.

sompylasar commented:
@dkorolev I'm not sure with "golden", I prefer to have a variable name that tells me what I'm dealing with, not for what purpose.

I've added the `-p` / `--prompt` command-line option to override the default prompt. The `--help` and `--version` were added for free by the `commander` module. Changed the run command in the README from `npm start` to `node run` because the former does not pass the command-line arguments to `node run`, and I prefer to have a single command to do a single thing (launch the REPL).

dkorolev commented:
LGTM, merging.
CC https://github.com/KnowSheet/KnowSheet-REPL/pull/8.
